mempool/cat/cache.go

```diff
@@ -0,0 +1,291 @@
+	c.list.Init()
+}
+
+func (c *LRUTxCache) Push(txKey types.TxKey) bool {
+	if c.staticSize == 0 {
+		return true
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	moved, ok := c.cacheMap[txKey]
+	if ok {
+		c.list.MoveToBack(moved)
+		return false
+	}
+
+	if c.list.Len() >= c.staticSize {
+		front := c.list.Front()
+		if front != nil {
+			frontKey := front.Value.(types.TxKey)
+			delete(c.cacheMap, frontKey)
+			c.list.Remove(front)
+		}
+	}
+
+	e := c.list.PushBack(txKey)
+	c.cacheMap[txKey] = e
+
+	return true
+}
+
+func (c *LRUTxCache) Remove(txKey types.TxKey) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	e := c.cacheMap[txKey]
+	delete(c.cacheMap, txKey)
+
+	if e != nil {
+		c.list.Remove(e)
+	}
+}
+
+func (c *LRUTxCache) Has(txKey types.TxKey) bool {
+	if c.staticSize == 0 {
+		return false
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	_, ok := c.cacheMap[txKey]
+	return ok
+}
+
+type EvictedTxInfo struct {
+	timeEvicted time.Time
+	priority    int64
+	gasWanted   int64
+	sender      string
+	size        int64
+}
+
+type EvictedTxCache struct {
+	staticSize int
+
+	mtx   tmsync.Mutex
+	cache map[types.TxKey]*EvictedTxInfo
+}
+
+func NewEvictedTxCache(size int) *EvictedTxCache {
+	return &EvictedTxCache{
+		staticSize: size,
+		cache:      make(map[types.TxKey]*EvictedTxInfo),
+	}
+}
+
+func (c *EvictedTxCache) Has(txKey types.TxKey) bool {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	_, exists := c.cache[txKey]
+	return exists
+}
+
+func (c *EvictedTxCache) Get(txKey types.TxKey) *EvictedTxInfo {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.cache[txKey]
+}
+
+func (c *EvictedTxCache) Push(wtx *wrappedTx) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	c.cache[wtx.key] = &EvictedTxInfo{
+		timeEvicted: time.Now().UTC(),
+		priority:    wtx.priority,
+		gasWanted:   wtx.gasWanted,
+		sender:      wtx.sender,
+		size:        wtx.size(),
+	}
+	// if cache too large, remove the oldest entry
+	if len(c.cache) > c.staticSize {
+		oldestTxKey := wtx.key
+		oldestTxTime := time.Now().UTC()
+		for key, info := range c.cache {
+			if info.timeEvicted.Before(oldestTxTime) {
+				oldestTxTime = info.timeEvicted
+				oldestTxKey = key
+			}
+		}
+		delete(c.cache, oldestTxKey)
+	}
+}
+
+func (c *EvictedTxCache) Pop(txKey types.TxKey) *EvictedTxInfo {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	info, exists := c.cache[txKey]
+	if !exists {
+		return nil
+	}
+	delete(c.cache, txKey)
+	return info
+}
+
+func (c *EvictedTxCache) Prune(limit time.Time) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	for key, info := range c.cache {
+		if info.timeEvicted.Before(limit) {
+			delete(c.cache, key)
+		}
+	}
+}
+
+func (c *EvictedTxCache) Reset() {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	c.cache = make(map[types.TxKey]*EvictedTxInfo)
+}
+
+// SeenTxSet records transactions that have been
+// seen by other peers but not yet by us
+type SeenTxSet struct {
+	mtx tmsync.Mutex
+	set map[types.TxKey]timestampedPeerSet
+}
+
+type timestampedPeerSet struct {
+	peers map[uint16]bool
```

Comment:
The `uint16` value maps to the index of a peer in the `peerMap` field of a `mempoolIDs`, right? If we could use the `p2p.ID` instead (as it seems to uniquely identify a peer),then we could avoid situations where there are two entires for the same peer in the peer set e.g., the connection of a peer drops and it re-connects.

Reply:
We should just clean up all ephemeral peer data like this whenever the connection drops.

If reconnection was a common occurrence we could look into keeping the mapping around for a period after the connection closed and thus giving the peer the same `uint16` if it were to reconnect.

Reply:
> We should just clean up all ephemeral peer data like this whenever the connection drops.

Right 

> If reconnection was a common occurrence we could look into keeping the mapping around for a period after the connection closed and thus giving the peer the same uint16 if it were to reconnect.

This works too, just need an extra book keeping for the index <> `p2p.ID` mapping (compared to using just `p2p.ID` and no index)